[#4566] Geoconnex reference collections

devincowan opened the pull request:
Resolves #4566 
<!--

Please read, and add your text at the bottom of this message.

Thanks for contributing code to HydroShare. In order to maintain code quality and expedite this process, please assist the development team by making sure the following is present in this pull request.

For more information, see https://docs.google.com/document/d/1dzxqlZW5fKNEyQSeKiSFq-SmS-VOPCva95XXkBjPExs

-->

### Pull Request Checklist: 
- [ ] Positive Test Case Written by Dev

<!-- Enter steps that a QA engineer, stakeholder, or user documentation writer would follow to test the positive or "successful" case of the functionality your code provides or fixes -->

- [ ] Automated Testing

<!-- Our Jenkins Instance is set up to automatically test every commit from a pull request. Code coverage must not decrease so new functionality or code paths added during a bug fix must have appropriate tests written. Every test must pass, including PEP8 code formatting tests. -->

- [ ] Sufficient User and Developer Documentation

<!-- Please email your positive test case lbrazil@cuahsi.org, who will make the decision regarding user documentation. -->

- [x] Passing Jenkins Build

<!-- Our Jenkins Instance is set up to automatically test every commit from a pull request. Every test must pass, including PEP8 code formatting tests. -->

- [x] Peer Code review and approval

<!-- This is the process by which a peer developer on the HydroShare team will read the changeset, provide feedback, and ultimately give a formal approval to the code before it passes PR status. -->

### Positive Test Case
1. [Enter positive test case here]

hydrocheck commented:
| METRIC | VALUE |
|:-----:|:-----:|

[https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-4737](https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-4737)

devincowan commented:
@Maurier the style question that I mentioned in our call today has to do with the buttons within this Geoconnex section.
You had suggested sticking with default Bootsrap buttons.
I tried that, but the scoped Vuetify styles within the Geoconnex app break the bootsrap button styles.
That's why I went with the Vuetify buttons:
<img width="1403" alt="image" src="https://user-images.githubusercontent.com/17934193/185959160-b11609fd-40dc-42d5-bb0a-eff2d3260825.png">

I recognize that this deviates substantially from the rest of the Resource Landing Page.
Do you think it is acceptable? Thanks...

Maurier reviewed (changes requested):
A few UI elements need to be adjusted for consistency.

![image](https://user-images.githubusercontent.com/2448568/185962793-0e666b90-2235-4141-8288-4492c44e89e1.png)
- Font should be smaller. Use `label` tag instead of `h4`.
- Make `info` icon gray. 
![image](https://user-images.githubusercontent.com/2448568/185963256-b2254402-d3a1-4e4f-8f98-6147faf26930.png)

![image](https://user-images.githubusercontent.com/2448568/185963416-78c8ed37-f3f5-4bc9-8c36-9bcbb63478b9.png)
- Move `info` icon outside of the button and place it on the right, next to it.
- The text in this button is too long. Consider consolidating into a shorter message and adding more content to the tooltip message. The button text should probably be 'Show map' and 'Search map'.
- Use similar existing casing for the button text. 'SHOW MAP' should be 'Show map'.
![image](https://user-images.githubusercontent.com/2448568/185964896-f5448cb8-2f8b-433d-9ce5-7b14184e11c9.png)
- Make sure layout is fully responsive.
- The text in the input items might need to be truncated with ellipsis ('...') in some cases, or maybe have them span multiple lines.
![image](https://user-images.githubusercontent.com/2448568/185967504-17a0f973-e6f4-4ac0-9a18-d652c26aeee2.png)
- Add space between buttons
- Use shorter button text.
- Move 'info' icons outside and place next to the button, on the right.
- Fix casing as described above.

Maurier commented:
@devincowan Loading some collections takes way too long. We need a way to make these requests simultaneously where possible. 

Loading these items took almost 10 minutes. 
![image](https://user-images.githubusercontent.com/2448568/185967080-be0be065-19be-4892-8a50-48228b6739bd.png)
![image](https://user-images.githubusercontent.com/2448568/185967291-8c863d37-4938-489d-9cc8-4073f70bc1b2.png)


We could probably improve this loading time with https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all

Maurier commented:
> @Maurier the style question that I mentioned in our call today has to do with the buttons within this Geoconnex section. You had suggested sticking with default Bootsrap buttons. I tried that, but the scoped Vuetify styles within the Geoconnex app break the bootsrap button styles. That's why I went with the Vuetify buttons: <img alt="image" width="1403" src="https://user-images.githubusercontent.com/17934193/185959160-b11609fd-40dc-42d5-bb0a-eff2d3260825.png">
> 
> I recognize that this deviates substantially from the rest of the Resource Landing Page. Do you think it is acceptable? Thanks...

I think we are going to need to resolve the scoping of CSS and use bootstrap buttons in order to be consistent.

horsburgh commented:
I'm with @Maurier here - the more consistent we can make the look and feel the more seamless it will be.

devincowan commented:
makes sense to me, thanks yall. I knew that in my head, but my heart was enticed by the slick look of those Vuetify spinners :)

devincowan commented:
@Maurier thank you for all of your catches and suggestions!
I believe that I have addressed them...
If you find these modifications acceptable, let me know and I will deploy somewhere public so that others can try it out.

devincowan commented on theme/static/js/hs-vue/geoconnex-reference-collection.js:
```diff
@@ -0,0 +1,1216 @@
+    this.setCustomItemRules();
+  },
+  async mounted() {
+    // TODO: artifacts / bits of text/elements are shown during initial load of page (less than 1 sec)
```
@Maurier a few weeks ago you mentioned that you might have a way to prevent the DOM artifacts that show briefly during page load. I haven't been able to figure it out...do I need to use a different Lifecycle hook? Thanks!

Maurier replied:
I'll have to try this out and look into it a little closer. Let me know if it's a good time to do so.

Maurier reviewed (changes requested):
@devincowan - Functionality and UX look good. I pushed some minor accessibility improvements and code cleanup. But commented out code remains that might need to be removed. I also saw a lot of cases where constants were declared using `let` instead of `const`.

Note that objects that are not meant to be reassigned entirely can be declared as constants. Objects declared as constants can still be mutated as needed.

Let me know when you've cleaned these up a bit and I'll approve.

devincowan commented:
thanks for the review @Maurier! I've cleaned things up a bit, let me know what you think...

Maurier reviewed (approved):
Looks good

devincowan commented:
@sblack-usu we are feeling pretty good about the front end for this new section on the Resource Landing Page.

I don't want to request a full review from you, but would you be willing to look at the following 2 diffs and tell me if my data model modifications make sense within the context of our other HS terms?

https://github.com/hydroshare/hydroshare/pull/4737/files#diff-3bc7182059367a2c6cf9e96428d08a0ce66f709baec5d00090abd19683582891

https://github.com/hydroshare/hydroshare/pull/4737/files#diff-f2f654aa015abc67926c60f43615a2375ceacb9f3c267a99b2b0a51770350c2d

We originally chose the DC term "Relation" for these related resources because we might eventually add collections that are not necessarily _geographically_ related.

sblack-usu commented:
@devincowan - we should talk about the RDF changes you made.  I'd like to know more about what is going on.

devincowan commented:
@Maurier re-requesting your approval on this PR. Things that have changed since your review:

- Fixed "info" tooltips so that they are viewable ON TOP of navigation
- Added the ability to "ignore" (prevent display) of certain specific features
- External URLs open in new tab
- Modified XML creation based on Scott's suggestions
- Shortened the info text/explanations
